keras_nlp/tokenizers/sentence_piece_tokenizer.py

```diff
@@ -0,0 +1,157 @@
+    and the [SentencePiece package](https://pypi.org/project/sentencepiece/).
+    The tokenization will run entirely within the Tensorflow graph, and can
+    be saved inside a `keras.Model`.
+
+    By default, the layer will output a `tf.RaggedTensor` where the last
+    dimension of the output is ragged after whitespace splitting and sub-word
+    tokenizing. If `sequence_length` is set, the layer will output a dense
+    `tf.Tensor` where all inputs have been padded or truncated to
+    `sequence_length`. The output dtype can be controlled via the `dtype`
+    argument, which should be either an integer or string type.
+
+    Args:
+        model_file: A path to a SentencePiece serialized model file. One
+            of `model_file` and `model_bytes` must be set.
+        model_bytes: A SentencePiece serialized model byte array. One of
+            `model_file` and `model_bytes` must be set.
+        sequence_length: If set, the output will be converted to a dense
+            tensor and padded/trimmed so all outputs are of sequence_length.
+
+    References:
+        - [Kudo and Richardson, 2018](https://arxiv.org/abs/1808.06226)
+
+    Examples:
+    ```python
+    # Train a SentencePiece vocabulary.
+    model = io.BytesIO()
+    ds = tf.data.Dataset.from_tensor_slices(["the quick brown fox."])
+    sentencepiece.SentencePieceTrainer.train(
+        sentence_iterator=ds.as_numpy_iterator(),
+        model_writer=model,
+        vocab_size=20,
+    )
+
+    # Tokenize inputs
+    tokenizer = SentencePieceTokenizer(model_bytes=model.getvalue())
+    ds = ds.map(tokenizer)
+    ```
+    """
+
+    def __init__(
+        self,
+        model_file: str = None,
+        model_bytes: bytes = None,
+        sequence_length: int = None,
+        **kwargs,
+    ) -> None:
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type or a string. "
+                    f"Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+
+        if model_file is None and model_bytes is None:
+            raise ValueError(
+                "One of `model_file` or `model_bytes` must be set. "
+                "Received: `model_file=None`, `model_bytes=None`."
+            )
+
+        if model_file is not None:
+            model_bytes = tf.io.gfile.GFile(model_file, "rb").read()
+        if isinstance(model_bytes, str):
+            model_bytes = base64.b64decode(model_bytes)
+
+        # Keras cannot serialize a bytestring, so we base64 encode the model
+        # byte array for saving.
+        self.model_bytes = base64.b64encode(model_bytes).decode("ascii")
```

Comment:
why do we apply `decode("ascii")` at encoding time, but not having it above?

Reply:
I'm confused at what you mean by above? We talked about this a bit earlier I think, this is only for serialization.